providers/amazon/src/airflow/providers/amazon/aws/bundles/s3.py

```diff
@@ -0,0 +1,203 @@
+                    raise e
+
+    def _download_s3_object_if_changed(self, s3_bucket, s3_object, local_target_path: Path):
+        should_download = False
+        download_msg = ""
+        if not local_target_path.exists():
+            should_download = True
+            download_msg = f"Local file {local_target_path} does not exist."
+        else:
+            local_stats = local_target_path.stat()
+
+            if s3_object.size != local_stats.st_size:
+                should_download = True
+                download_msg = (
+                    f"S3 object size ({s3_object.size}) and local file size ({local_stats.st_size}) differ."
+                )
+
+            s3_last_modified = s3_object.last_modified
+            if local_stats.st_mtime < s3_last_modified.microsecond:
+                should_download = True
+                download_msg = f"S3 object last modified ({s3_last_modified.microsecond}) and local file last modified ({local_stats.st_mtime}) differ."
+
+        if should_download:
+            s3_bucket.download_file(s3_object.key, local_target_path)
+            self.log.debug(
+                "%s Downloaded %s to %s", download_msg, s3_object.key, local_target_path.as_posix()
+            )
+        else:
+            self.log.debug(
+                "Local file %s is up-to-date with S3 object %s. Skipping download.",
+                local_target_path.as_posix(),
+                s3_object.key,
+            )
+
+    def _download_s3_dags(self):
+        """Download DAG files from the S3 bucket to the local directory."""
+        self.log.debug(
+            "Downloading DAGs from s3://%s/%s to %s", self.bucket_name, self.prefix, self.s3_dags_dir
+        )
+        local_s3_objects = []
+        s3_bucket = self.s3_hook.get_bucket(self.bucket_name)
+        for obj in s3_bucket.objects.filter(Prefix=self.prefix):
+            obj_path = Path(obj.key)
+            local_target_path = self.s3_dags_dir.joinpath(obj_path.relative_to(self.prefix))
+            if not local_target_path.parent.exists():
+                local_target_path.parent.mkdir(parents=True, exist_ok=True)
+                self.log.debug("Created local directory: %s", local_target_path.parent)
+            self._download_s3_object_if_changed(
+                s3_bucket=s3_bucket, s3_object=obj, local_target_path=local_target_path
+            )
+            local_s3_objects.append(local_target_path)
+
+        self._delete_stale_local_files(current_s3_objects=local_s3_objects)
+
+    def __repr__(self):
+        return (
+            f"<S3DagBundle("
+            f"name={self.name!r}, "
+            f"bucket_name={self.bucket_name!r}, "
+            f"prefix={self.prefix!r}, "
+            f"version={self.version!r}"
+            f")>"
+        )
+
+    def get_current_version(self) -> str | None:
+        """Return the current version of the DAG bundle. Currently not supported."""
+        return None
```

Comment:
Any reason we don't support versioning here?

Reply:
not sure how this could be done with S3. so left it out for the initial implementation.

Reply:
I think there is bucket versioning, but yeah, we can explore that after this initial implementation